internal/worker/apiaddresssetter/worker.go

```diff
@@ -0,0 +1,366 @@
+	}
+	if config.Logger == nil {
+		return errors.New("nil Logger not valid").Add(coreerrors.NotValid)
+	}
+	return nil
+}
+
+// New returns a new worker that maintains the api addresses for the controller
+// nodes.
+func New(config Config) (worker.Worker, error) {
+	if err := config.Validate(); err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	w := &apiAddressSetterWorker{
+		config:                config,
+		controllerNodeChanges: make(chan struct{}),
+		controllerTrackers:    make(map[string]*controllerTracker),
+		runner: worker.NewRunner(worker.RunnerParams{
+			IsFatal:       func(error) bool { return false },
+			ShouldRestart: func(error) bool { return false },
+			Logger:        internalworker.WrapLogger(config.Logger),
+		}),
+	}
+	err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+		Init: []worker.Worker{w.runner},
+	})
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+	return w, nil
+}
+
+// Kill is part of the worker.Worker interface.
+func (w *apiAddressSetterWorker) Kill() {
+	w.catacomb.Kill(nil)
+}
+
+// Wait is part of the worker.Worker interface.
+func (w *apiAddressSetterWorker) Wait() error {
+	return w.catacomb.Wait()
+}
+
+func (w *apiAddressSetterWorker) loop() error {
+	controllerNodeChanges, err := w.watchForControllerNodeChanges()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	ctx, cancel := w.scopedContext()
+	defer cancel()
+
+	configChanges, err := w.watchForConfigChanges(ctx)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	for {
+		w.config.Logger.Tracef(ctx, "waiting for controller nodes or addresses changes")
+		select {
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+		case <-controllerNodeChanges:
+			// A controller was added or removed.
+			w.config.Logger.Tracef(ctx, "<-controllerChanges")
+			changed, err := w.updateControllerNodes(ctx)
+			if err != nil {
+				return errors.Capture(err)
+			}
+			if !changed {
+				continue
+			}
+			w.config.Logger.Tracef(ctx, "controller node added or removed")
+		case <-w.controllerNodeChanges:
+			// One of the controller nodes changed.
+			w.config.Logger.Tracef(ctx, "<-w.controllerChanges")
+		case <-configChanges:
+			// Controller config has changed.
+			w.config.Logger.Tracef(ctx, "<-w.configChanges")
+
+			// If a config change wakes up the loop before the topology has
+			// been represented in the worker's controller trackers, ignore it;
+			// errors will occur when trying to determine peer group changes.
+			// Continuing is OK because subsequent invocations of the loop will
+			// pick up the most recent config from state anyway.
+			if len(w.controllerTrackers) == 0 {
+				w.config.Logger.Tracef(ctx, "no controller information, ignoring config change")
+				continue
+			}
+		}
+
+		if err := w.updateAPIAddresses(ctx); err != nil {
+			w.config.Logger.Errorf(ctx, "cannot update api addresses: %v", err)
+		}
+	}
```

Comment:
Have you considered the case where the watcher restarts and loses it's internal state? 

Is that why we don't consume the initial event for controller nodes?

Reply:
Indeed, the trackers are created every time we have changes including the initial event, so the internal map (state) is resilient to the worker restarts.